### PR TITLE
stable ellipse fitting

### PR DESCRIPTION
**Pull request by Borda**
## Description
Fix for ellipse fitting is not stable #2379

## Checklist
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
This method of least Squares fitting for Ellipses is taken from the paper by Halir, R., and Flusser, J. [Numerically Stable Direct Least Squares Fitting of Ellipses](http://autotrace.sourceforge.net/WSCG98.pdf)

### CODE REVIEW

**Comment by Borda** [comment]
waiting for #2395

**Comment by Borda** [comment]
I have Account at Travis now, so how can I rerun a build?
[Travis - scikit-image/scikit-image](https://travis-ci.org/scikit-image/scikit-image/builds/182043501)
for some other repositories I have an option as rebuild, but no here 
[Travis - Borda/scikit-image](https://travis-ci.org/Borda/scikit-image)

**Comment by sciunto** [comment]
@Borda You need to rebase on master, such that the fix for travis in included in your branch. See .github/CONTRIBUTING.txt for guidance.

**Comment by sciunto** [comment]
OK, looks like you rebased, the test passes.

**Comment by emmanuelle** [comment]
@Borda, this PR is work in progress, isn't it?

**Comment by Borda** [comment]
correct, I hope I can do it in following days... also I was hoping to start this implementations after merging #2371 which could be in comfit.

**Comment by Borda** [comment]
Does anyone have an advice how to debug the test on the CI results, because I run the same data locally with always the same results but much different from the CI results...
File "skimage\measure\tests\test_fit.py", line 218, in test_ellipse_model_estimate_from_data

my results are: `array([  4.98107e+02,   8.40229e+02,   1.35715e+02, 1.63236e+02])`
CI results are: `array([  419.895785,   628.612303,  1970.032086,   244.014599])`

note, the main difference is in the a and b parameter (3. and 4. element)

**Comment by Borda** [comment]
@emmanuelle actually the implementation is done now, just wondering why the test fails on Appveyor, maybe a number sounding... Should I slightly change the newly added test?

File: skimage/measure/fit.py [hunk 1]
```diff
@@ -457,31 +456,28 @@ class EllipseModel(BaseModel):
...
-    ellipse::
-
-        min{ sum(d_i**2) } = min{ sum((x_i - xt)**2 + (y_i - yt)**2) }
-
-    Thus you have ``2 * N`` equations (x_i, y_i) for ``N + 5`` unknowns (t_i,
-    xc, yc, a, b, theta), which gives you an effective redundancy of ``N - 5``.
+    This estimator minimizes
+    The estimator is based on a least squares minimization. The optimal
+    solution is computed directly, no iterations are required. This leads
+    to a simple, stable and robust fitting method.
```
Comment (jni) [inline skimage/measure/fit.py]: @Borda this docstring needs fixing

File: skimage/measure/fit.py [hunk 2]
```diff
@@ -457,31 +456,28 @@ class EllipseModel(BaseModel):
...
 
-    A minimum number of 5 points is required to solve for the parameters.
-
     Attributes
     ----------
     params : tuple
-        Ellipse model parameters in the following order `xc`, `yc`, `a`,
-        `b`, `theta`.
+        Ellipse model parameters in the following order `xc`, `yc`, `a`, `b`,
+        `theta`.
```
Comment (jni) [inline skimage/measure/fit.py]: It looks like the `params` attribute will be the output of `nan_to_num`, which is a NumPy array, not a tuple?

File: skimage/measure/fit.py [hunk 3]
```diff
@@ -492,66 +488,76 @@ def estimate(self, data):
         success : bool
             True, if model estimation succeeds.
 
-        """
 
+        References:
+        -----------
+            [1] Halir, R., Flusser, J.: 'Numerically Stable Direct Least Squares
+                Fitting of Ellipses'
```
Comment (jni) [inline skimage/measure/fit.py]: Use `.. [1] Halir, R.` etc., with no indent relative to the heading. See other docstrings in the library for examples.

File: skimage/measure/fit.py [hunk 4]
```diff
@@ -492,66 +488,76 @@ def estimate(self, data):
...
-        r0 = np.sqrt((x - xc0)**2 + (y - yc0)**2).mean()
-        params0[:5] = (xc0, yc0, r0, 0, 0)
-        params0[5:] = np.arctan2(y - yc0, x - xc0)
-
-        params, _ = optimize.leastsq(fun, params0, Dfun=Dfun, col_deriv=True)
-
-        self.params = params[:5]
-
+        # Quadratic part of design matrix [eqn. 15] from [1]
+        D1 = np.mat(np.vstack([x ** 2, x * y, y ** 2])).T
```
Comment (jni) [inline skimage/measure/fit.py]: We discourage the use of `numpy.matrix` objects at scikit-image. (I might be mistaken on this but if that's the case, someone else should jump in and correct me.) The reason is that these objects are very similar but critically different to the far more common numpy arrays. Therefore, a reader seeing a matrix object being used might incorrectly think it's an array, and incorrectly interpret the use of e.g. the `*` operator.

I know it's uglier, but please change the code to use numpy arrays everywhere and `np.dot` for matrix multiplication. (At some indeterminate point in the future, we can move to Python 3.5+ and use the `@` operator to get both unambiguity and elegance.)

Some other devs might want to jump in and opine here... @emmanuelle @stefanv @ahojnnes ?

File: skimage/measure/fit.py [hunk 5]
```diff
@@ -811,23 +815,23 @@ def ransac(data, model_class, min_samples, residual_threshold,
     >>> model = EllipseModel()
     >>> model.estimate(data)
     True
-    >>> model.params # doctest: +SKIP
-    array([ -3.30354146e+03,  -2.87791160e+03,   5.59062118e+03,
-             7.84365066e+00,   7.19203152e-01])
-
+    >>> np.round(model.params) # doctest: +SKIP
```
Comment (jni) [inline skimage/measure/fit.py]: Two spaces before `#`

File: skimage/measure/fit.py (hunk 3 above)
Comment (sciunto) [inline skimage/measure/fit.py]: Please, add the DOI. Format: `DOI:xxxxxxx`

**Comment by Borda** [comment]
@jni I am not sure what you mean that I solve it in Travis before, Once Travis crashed just in the beginning and later (another day) it worked fine. But here it gives me the strange result always - several times in week period. The problem is that I am not able to reproduce it to find the source.

**Comment by jni** [comment]
@Borda Ah, sorry, I misread your earlier comments. Hmmm... Well, I'm fresh out of ideas!

File: skimage/measure/fit.py (hunk 3 above)
Comment (Borda) [inline skimage/measure/fit.py]: sorry, I cannot find DOI... [Numerically Stable Direct Least Squares Fitting Of Ellipses](https://www.researchgate.net/publication/2820897_Numerically_Stable_Direct_Least_Squares_Fitting_Of_Ellipses)

File: skimage/measure/fit.py (hunk 2 above)
Comment (Borda) [inline skimage/measure/fit.py]: it worked with assignment from np.array before, but now it is changed to tuple...

**Comment by Borda** [comment]
@jni is it fine now?

File: skimage/measure/fit.py [hunk 6]
```diff
@@ -457,31 +456,27 @@ class EllipseModel(BaseModel):
...
-        `b`, `theta`.
+        Ellipse model parameters in the following order `xc`, `yc`, `a`, `b`,
+        `theta`.
 
     """
 
     def estimate(self, data):
         """Estimate circle model from data using total least squares.
 
+        Original Implementation: Ben Hammel, Nick Sullivan-Molina
```
Comment (ahojnnes) [inline skimage/measure/fit.py]: I would change this to be a comment in the code rather than the doc string.

File: skimage/measure/fit.py [hunk 7]
```diff
@@ -492,66 +487,78 @@ def estimate(self, data):
...
-        r0 = np.sqrt((x - xc0)**2 + (y - yc0)**2).mean()
-        params0[:5] = (xc0, yc0, r0, 0, 0)
-        params0[5:] = np.arctan2(y - yc0, x - xc0)
-
-        params, _ = optimize.leastsq(fun, params0, Dfun=Dfun, col_deriv=True)
-
-        self.params = params[:5]
-
+        # Quadratic part of design matrix [eqn. 15] from [1]
+        D1 = np.vstack([x ** 2, x * y, y ** 2]).T
```
Comment (ahojnnes) [inline skimage/measure/fit.py]: replace with `np.hstack` and remove the `.T`?

File: skimage/measure/fit.py [hunk 8]
```diff
@@ -492,66 +487,78 @@ def estimate(self, data):
...
+        # Linear part of design matrix [eqn. 16] from [1]
+        D2 = np.vstack([x, y, np.ones(len(x))]).T
```
Comment (ahojnnes) [inline skimage/measure/fit.py]: Same here...

File: skimage/measure/fit.py [hunk 9]
```diff
@@ -492,66 +487,78 @@ def estimate(self, data):
...
+
+        # Constraint matrix [eqn. 18]
+        C1 = np.array([[0., 0., 2.], [0., -1., 0.], [2., 0., 0.]])
+
+        # Reduced scatter matrix [eqn. 29]
+        M = np.linalg.inv(C1).dot(S1 - np.dot(S2, np.linalg.inv(S3)).dot(S2.T))
+
+        # M*|a b c >=l|a b c >. Find eigenvalues and eigenvectors
+        # from this equation [eqn. 28]
+        eval, evec = np.linalg.eig(M)
```
Comment (ahojnnes) [inline skimage/measure/fit.py]: `eval` is an unfortunate name in Python. How about `eig_vals` and `eig_vecs`?

File: skimage/measure/fit.py (hunk 7 above)
Comment (Borda) [inline skimage/measure/fit.py]: not really, it this usage the `np.vstack` gives 3 x N and the `np.vstack` gives 1 x 3*N

**Comment by jni** [comment]
@Borda LOL your change to the test is a little Volkswagen-y. =P I'm a little concerned that we don't understand the cause of the test failure, but I'm happy enough to ignore it.

**Comment by Borda** [comment]
@jni I am not happy about it also but so far I cannot trace it what really happening in the CI, I do not know a way how to fix it... on the other hand the previous estimation fails in order 1e7 so now with the result difference only something less then 10 times it is quite fine... :)

**Comment by jni** [comment]
@Borda that was my intuition as well. =) Thanks for the contribution! Just waiting for @ahojnnes to approve.

**Comment by Borda** [comment]
@jni you are welcome... but now it seems that the building doc is failing...

```
make[1]: *** [html] Error 1
make[1]: Leaving directory `/home/travis/build/scikit-image/scikit-image/doc'
make: *** [html] Error 2
```

**Comment by ahojnnes** [comment]
LGTM, waiting for the tests to pass :-)

**Comment by Borda** [comment]
@jni @ahojnnes done :)

**Comment by ahojnnes** [comment]
@Borda Thanks for the contribution!

**Pull request merged**